example_notebooks/cube_4d.ipynb

```diff
@@ -0,0 +1,354 @@
+    "            self.line_plot.set_xlabel(self.plot_xlabel)\n",
+    "            self.line_plot.set_ylabel(self.plot_ylabel)\n",
+    "\n",
+    "            clear_output(wait=True)\n",
+    "            display(self.line_plot.figure)\n",
+    "\n",
+    "    def show_slice(self, n):\n",
+    "        image = self._viewer.get_image()\n",
+    "        image.set_naxispath([n, self._4d_idx])\n",
+    "        self._viewer.redraw(whence=0)\n",
+    "        self._cur_islice = n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "logger = get_logger('my viewer', log_stderr=True, log_file=None, level=30)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "w = CubeWidget(logger=logger)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Generate a fake 4D data cube."
+   ]
+  },
+  {
+   "cell_type": "raw",
+   "metadata": {},
+   "source": [
+    "import numpy as np\n",
+    "\n",
+    "pf = None  # Not using FITS file pointer\n",
+    "cube4d = np.random.random((2, 5, 100, 200))\n",
+    "\n",
+    "# Convert to NDData.\n",
+    "im = NDData(cube4d, wcs=WCS(naxis=4))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**--- OR ---**\n",
+    "\n",
+    "Load an existing data cube.\n",
+    "\n",
+    "Dask support requires https://github.com/ejeschke/ginga/pull/805."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import dask.array as da\n",
+    "from astropy.io import fits\n",
+    "\n",
+    "# This is a large cube that is 6GB.\n",
+    "filename = '/redkeep/ironthrone/ssb/stginga/test_data/NRCN815A_LIN_20160115_uncal.fits'\n",
+    "pf = fits.open(filename, memmap=True)\n",
+    "\n",
+    "# There is no WCS assigned for UNCAL file yet.\n",
+    "# This is a dummy WCS.\n",
+    "im_wcs = WCS(pf[1].header)\n",
+    "\n",
+    "# Convert to NDData that wraps a Dask array.\n",
+    "# Must provide Dask with a name to avoid hashing.\n",
+    "# Shape is (1, 190, 2048, 2048)\n",
+    "im = NDData(da.from_array(pf[1].data, name='nircam_sci'), wcs=im_wcs)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Grab the `z`-dimension for slider."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_3d = im.data.shape[1]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Load `NDData` into viewer widget."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Loading a fake WCS into widget sometimes gives error.\n",
+    "# Try reloading before giving up.\n",
+    "w.load_nddata(im)"
```

Comment:
The first time I call this, I get this error. But when I rerun the cell subsequently, it goes way. @ejeschke , do you know why?

```
2020-02-07 13:05:18,025 | E | Callback.py:172 (_do_callbacks) |
  Error making callback 'cursor-changed': DEC (2048.500000) > 90.0
2020-02-07 13:05:18,026 | E | Callback.py:173 (_do_callbacks) | Traceback:
  File ".../ginga/misc/Callback.py", line 156, in _do_callbacks
    res = method(*cb_args, **cb_kwargs)

  File ".../astrowidgets/core.py", line 214, in _mouse_move_cb
    raDegToString(ra), decDegToString(dec))

  File ".../ginga/util/wcs.py", line 137, in decDegToString
    sign, dec_degree, dec_min, dec_sec = degToDms(dec_deg)

  File ".../ginga/util/wcs.py", line 81, in degToDms
    assert dec <= 90, WCSError("DEC (%f) > 90.0" % (dec))
```

Reply:
That's a WCS check that we probably should look at.  It's just trying to format the coordinates to string values and finding an invalid DEC.  Even though this is a load operation, the formatting happens because of a callback in the viewer when the image changes.  Not sure why it wouldn't occur again if you re-execute the cell, but it might depend on where it thinks the cursor is WRT the window.